### PR TITLE
fix(ios): send boolean event properties as JSON booleans

### LINKED ISSUE #68: iOS: Boolean event properties are captured as 0/1 instead of true/false
### Version

0.3.0 (observed live). The capture path at current `main` passes properties through unchanged as well (see analysis below), so it should reproduce on 0.5.0 too.

### Platform(s) affected

- [ ] Android
- [x] iOS
- [ ] JS

### Steps to Reproduce

1. From `commonMain`, capture an event with a `Boolean` property value:

   ```kotlin
   PostHog.capture("Test Event", properties = mapOf("my_flag" to true))
   ```

2. Run the app on iOS and let the event flush.
3. Inspect the event's properties in PostHog.

### Expected Result

`my_flag` arrives as a JSON boolean (`true`/`false`), matching what the same common code produces on Android.

### Actual Result

On iOS the property arrives as a number: `1` (or `0` for `false`). Android correctly sends `true`/`false`. An app shipping the same common code on both platforms therefore ends up with mixed value types on a single PostHog property, which breaks boolean property filters and splits insight breakdowns into `true`/`false`/`1`/`0` buckets.

### Analysis

`platformCapture` in `PostHog.ios.kt` hands the Kotlin `Map<String, Any>` straight across the Kotlin/Native ↔ Objective-C boundary without any value conversion:

- On 0.3.0: `properties as? Map<Any?, *>` → `PostHogBridge.capture(...)`, which forwards the `NSDictionary` unchanged to `PostHogSDK.shared.capture`.
- On current `main`: `properties as? Map<Any?, *>` → `PostHogSDK.shared.captureWithEvent(...)` directly.

When a boxed Kotlin `Boolean` crosses that boundary inside a generic collection, Kotlin/Native bridges it as `KotlinBoolean` — an `NSNumber` subclass provided by the Kotlin/Native runtime, which is **not** one of the `CFBoolean` singletons (`kCFBooleanTrue`/`kCFBooleanFalse`) that `NSNumber(value: Bool)` returns. `JSONSerialization` only emits `true`/`false` for actual `CFBoolean` instances; every other `NSNumber` is serialized numerically. So the boolean hits the wire as `1`/`0`.

(The JVM path is unaffected because the value stays a `java.lang.Boolean` all the way to serialization.)

### Suggested fix

Sanitize property values on the iOS side before handing them to posthog-ios: recursively map boxed Kotlin `Boolean` values (including inside nested maps/lists) to `NSNumber.numberWithBool()`, which returns the `CFBoolean` singletons and serializes as JSON `true`/`false`.

App-side workaround in the meantime: an `expect`/`actual` property sanitizer that performs the same `Boolean` → `NSNumber.numberWithBool()` conversion in `iosMain` before calling `PostHog.capture`.

### PR DESCRIPTION
**Pull request by turnipdabeets**
## :bulb: Motivation and Context

Fixes #68.

Boolean event properties captured from `commonMain` arrive in PostHog as `1`/`0` on iOS while the same common code sends `true`/`false` on Android, so a single property ends up with mixed value types — boolean filters break and insight breakdowns split across `true`/`false`/`1`/`0`.

Kotlin/Native bridges a boxed Kotlin `Boolean` as `KotlinBoolean`, an `NSNumber` subclass that is **not** one of the `CFBoolean` singletons. `JSONSerialization` writes `true`/`false` only for `CFBoolean` and serializes every other `NSNumber` numerically, so the boolean hits the wire as a number.

Two things worth flagging beyond the issue report:

**1. The workaround suggested in the issue does not work.** `NSNumber.numberWithBool()` called *from Kotlin* returns a value that Kotlin/Native converts straight back to a Kotlin `Boolean`, which is then re-boxed as `KotlinBoolean` on the way out. The same is true for `kCFBooleanTrue` and for a boolean parsed out of JSON: a `CFBoolean` simply cannot be held in a Kotlin variable. The containers therefore have to be assembled on the Objective-C side, which is what `toNativeProperties()` does — it never reads the boolean values back.

**2. The bug also corrupted properties the native SDK sets itself.** `processBeforeSend` is registered unconditionally (to stamp `$lib`/`$lib_version`) and rewrote the *entire* property dictionary through Kotlin on every event, after `buildProperties` had already run. That re-boxed nine native booleans on every single event. Baseline payload from `main`:

```json
"$is_testflight":0,"$is_identified":0,"$process_person_profile":0,"$is_emulator":1,
"$network_wifi":1,"$network_cellular":0,"$is_sideloaded":0,
"$is_ios_running_on_mac":0,"$is_mac_catalyst_app":0
```

When no user `beforeSend` callback is configured, the SDK metadata is now stamped onto the native dictionary instead of round-tripping it through Kotlin, so those values are never converted. That path is also cheaper than what it replaces.

`platformRegister` is normalized separately because super properties are persisted to disk as JSON — a boolean re-boxed there stays a number permanently, and no `beforeSend` fix can recover it.

## Upgrade impact

Kept out of the change intent so the changelog line stays scannable. Traced through the PostHog source, and the filter behaviour was confirmed by evaluating the compiled expressions against a live project rather than reasoning about them.

- **Old and new values cannot both match one filter.** A property definition's type is inferred once and never re-inferred — the upsert sets `property_type` only `WHERE property_type IS NULL` (`rust/property-defs-rs/src/batch_ingestion.rs:663-667`). On a `Boolean` definition a stored `1` compiles to `toBool(transform(toString(prop), ['true','false'], [1,0], NULL))`, which evaluates to `NULL`; on a `Numeric` definition a stored `true` compiles to `toFloat(prop)`, also `NULL`, which additionally makes those events read as **is not set**. The left-hand side is NULL either way, so listing both `true` and `1` does not help. A project that was iOS-only has these properties typed `Numeric` and needs the type changed to Boolean under Data management → Properties. A project that also ships Android already has them typed `Boolean`, so its iOS events simply start matching filters that never worked before.
- **Only event properties age out.** Person properties (`identify`, `setPersonProperties`), group properties (`group`) and super properties (`register`) persist until they are set again — a dormant user keeps `1` on their person record indefinitely. Super properties are additionally stored on disk as JSON, so they keep going out as numbers until `register` is called again. A `$set_once` boolean written as `1` can never be corrected, since the check is presence-only (`person-update.ts:162`); it needs an explicit `$set` backfill. Plain `$set` self-heals on the next write, because `1 !== true`.
- **Anonymous iOS events stop creating person profiles.** `$process_person_profile` is one of the booleans the old round-trip corrupted, and ingestion compares it strictly against `false` (`person-utils.ts:16-30`), so `0` meant *create the profile* — overriding `personProfiles = identifiedOnly`, which is the KMP default. After this fix, person counts stop growing from anonymous iOS traffic, the sampled `invalid_process_person_profile` ingestion warnings stop, and the enhanced-persons usage metric drops, since it counts only `person_mode IN ('full', 'force_upgrade')`. This was invisible in the data because ingestion deletes the key before storing the event.
- Null-valued properties the native SDK sets, such as `$feature_flag_response` for an absent flag, now reach the wire as `null` instead of being dropped. Cosmetic for filtering: present-but-null is treated exactly like absent, confirmed by evaluating `nullIf(nullIf(JSONExtractRaw(...), ''), 'null')` and `is set` against a live project.

## :green_heart: How did you test it?

**Unit tests** — new `NativePropertiesTest` in `iosTest`, asserting on the JSON text `JSONSerialization` produces (the values themselves are converted back on the way into Kotlin, so they can't be asserted on directly): top-level booleans, booleans nested in maps and lists, other value types left unchanged (`Byte` in particular must not become a boolean), and the metadata stamp preserving native booleans.

`iosTest` was never wired into CI — the iOS job only compiled. This PR adds `:posthog-kmp:iosSimulatorArm64Test` to it, so the new tests and the two existing iOS test classes now actually run. 27 tests pass locally on `iosSimulatorArm64`.

**End-to-end** — a throwaway test (not committed) set up the SDK, captured `mapOf("my_flag" to true, "off_flag" to false, "count" to 7)`, and read the queued event straight off disk. Same harness, before and after:

| | payload |
|---|---|
| `main` | `"my_flag":1,"off_flag":0,"count":7` plus the nine native booleans as `0`/`1` |
| this branch | `"my_flag":true,"off_flag":false,"count":7` plus `"$is_identified":false,"$is_emulator":true,"$network_wifi":true,…` |

Also run: `detekt` clean, `compileKotlinIosArm64` / `compileKotlinIosX64` with no new warnings, `jvmTest` green.

**Review pass.** Two independent reviews were run over this branch, each re-verifying the load-bearing claim by compiling and running standalone Kotlin/Native programs rather than taking it on faith. Both confirmed that `NSNumber.numberWithBool()`, `kCFBooleanTrue` via `interpretObjCPointer`, and a boolean read out of parsed JSON all come back as `kotlin.Boolean` and re-box as `1` — so the Objective-C-side assembly here is the only construction that survives, and the workaround suggested in #68 does not work. Two findings were real and are fixed in the second commit:

- **Null-valued properties.** posthog-ios sets `$feature_flag_response`, `$feature_flag_reason`, `$feature_flag_id` and `$feature_flag_version` to `NSNull` for an absent flag. `NSNull` reads back into Kotlin as `null`, so `filterValues { it != null }` dropped those keys. The metadata fast path kept them while the before-send path stripped them, which made event contents depend on whether the app registered a callback. Both paths keep them now.
- **Unbounded recursion.** A self-referential property value crashed the host app, where posthog-ios previously dropped the key and logged, because the conversion now runs before its `isValidObject` guard. Verified: uncapped `exit=139`, capped `exit=0`. Nesting past 64 levels is handed over unconverted for that guard to reject, as it did before.

A second review pass then caught that the depth bound I'd added was itself miscalibrated: `JSONSerialization` accepts 511 container levels, so between 65 and 511 a payload stayed serializable while the booleans inside it silently reverted to `1`/`0` — reachable through session replay, where wireframe trees nest two levels per view and carry `disabled`/`checked` booleans. The bound now tracks Foundation's, verified at that limit on a secondary thread (where the native SDK may run before-send callbacks): `valid=true keepsBoolean=true`. That pass also removed a test that duplicated an existing case while appearing to cover the null-property change, and replaced assertions that were really testing Foundation's depth limit with one that pins this bound.

Findings raised and retracted after falsification: `dictionaryWithObjects:forKeys:` raising on a nil key (a Kotlin `null` key bridges to `NSNull`, never nil), the dropped `filterKeys { it is String }` (native properties are `[String: Any]`), and `trueBox`/`falseBox` thread-safety (initialized once, and the `CFBoolean`s are process-wide singletons either way).

Not tested: a real device, and a user-supplied `beforeSend` callback end to end (the writeback path it takes is covered by the unit tests). `processBeforeSend` has no direct test because posthog-ios's `PostHogEvent` initializer is not public, so a native event cannot be constructed from Kotlin; the testable half is covered through `withSdkMetadata`.

**Session replay:** snapshots are generated on the main thread (`PostHogReplayIntegration.swift:1372`) and throttled to one a second, so rebuilding their property tree there would have been a stall repeating once a second for apps that also register a `beforeSend` callback. The writeback now compares top-level values by identity and rebuilds only what the callback actually replaced; everything else carries over as the native objects it already is. Independently measured on an 8,845-node replay payload: **1884 µs → 1.9 µs** when the callback leaves properties alone, 2.6 µs with one property replaced. A callback that replaces every value costs 3.5% over the old path, so the optimization never meaningfully regresses.

That optimization was reviewed separately, since it landed after the two earlier rounds. The review found two defects in it, both fixed in `9c895a5`: a property *added* with a `null` value was dropped (an added key reads back as `null` from the map handed to the callback, so it compared equal to its own absence), and the SDK metadata was stamped after the callback's removals, making `$lib` removal a silent no-op while overwriting it still worked — incoherent, and out of parity with JS and JVM which both honour removal. The diff was also extracted from a `private` function taking a native event type Kotlin cannot construct, so it is now covered by tests that fail against the pre-fix implementation.

A final round covered the review feedback on this PR and a fresh pass over the branch. Three changes came out of it:

- **Cycle detection (`4ba7206`).** The depth bound stopped a self-reference that recursed in a straight line, but not one that branches: a container holding itself twice is copied `2^512` times before the bound can return, so capture pegs the CPU instead of letting posthog-ios reject the value. The conversion now tracks the containers it is inside and hands a repeat back uncopied, exactly as the depth bound does. Verified in both directions — the new test does not finish within 300s against the pre-fix source, and passes in about a second after. The bound stays as the stack guard for legitimately deep acyclic payloads; Foundation's real write limit was measured at 511 levels, so `MAX_DEPTH = 512` sits on the safe side.
- **Callback nulls (`ef930aa`).** Dropping null values on the way out was lost when the writeback became a diff, so a callback setting a property to `null` started emitting an explicit `null` on iOS while the same shared-code callback still removed the key on Android and the JVM. Nulling a property removes it again. Properties the native SDK set to `NSNull` are untouched, since the callback hands those back unchanged and they never enter the diff. The logic lives in `propertyDelta` rather than `processBeforeSend`, because the latter takes a native event type Kotlin cannot construct and so cannot be tested.
- **Change intent (`17e6c98`, `1413154`, `d751bfb`).** Rewritten twice as the upgrade impact above was actually verified: first because the stale values are not limited to events and do not age out, then because telling users to match both `true` and `1` is not advice that works on a typed property.

Neither reviewer found a blocker in the cycle-detection change itself. The adversarial pass confirmed the push/pop balance, that there is no off-by-one against the old `depth` parameter, that identity comparison cannot produce a false positive, and that a detected cycle is still dropped and logged by posthog-ios's `sanitizeDictionary` rather than slipping through to `JSONSerialization`.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [ ] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

### If releasing new changes

- [x] Ran `pnpm change` to generate a change intent file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

Written with Claude Code (Opus 5). The investigation ran first as `/investigate-github-issue` on #68 and confirmed the report against `main`; the root-cause claim was then verified empirically rather than from reasoning, by compiling small Kotlin/Native programs against `kotlinc-native` and serializing the exact map shapes that cross the boundary.

Approaches tried and rejected along the way:

- **A Swift shim in the SwiftPM bridge**, which is where a normalizer would naturally live. Rejected: `PostHogBridge.swift` was deleted in #65 when the build moved to Kotlin's built-in `swiftPMDependencies`, and re-adding a local Swift package to a build that had just been simplified to remove one seemed like the wrong trade for this bug.
- **Holding a `CFBoolean` in Kotlin** via `NSNumber.numberWithBool`, `interpretObjCPointer(kCFBooleanTrue)`, or a JSON round trip. All three measurably come back as Kotlin `Boolean` and re-box — this is also why the workaround in the issue doesn't help.
- **A CoreFoundation route** (`CFDictionaryCreateMutable` + `CFDictionarySetValue` + toll-free bridge). It works, but needs manual `CFRelease`/`interpretObjCPointer` handling; the Foundation-only route in this PR is equivalent and has no lifetime concerns.
- **Normalizing everything at the `beforeSend` writeback**, which would have been a one-line fix covering all surfaces. Rejected because it deep-walks every event's properties, including large `$snapshot` payloads for session-replay users.

One judgement call worth a reviewer's eye: `withSdkMetadata` is `internal` rather than `private` so `iosTest` can assert on it.

### CODE REVIEW

File: posthog-kmp/src/iosMain/kotlin/com/posthog/kmp/NativeProperties.ios.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+private fun NSMutableArray.appendNative(value: Any?, depth: Int) {
+    when {
+        value == null -> addObject(NSNull())
+        value is Boolean -> {
+            // The fallback only restores the 1/0 behaviour; it never fails.
+            val box = if (value) trueBox else falseBox
+            if (box != null) addObjectsFromArray(box) else addObject(value)
+        }
+        depth >= MAX_DEPTH -> addObject(value)
```
Comment (marandaneto) [inline posthog-kmp/src/iosMain/kotlin/com/posthog/kmp/NativeProperties.ios.kt]: blocking: Detect cycles rather than only bounding depth — A container that references itself through multiple entries branches at every level here, so the current depth-only guard performs roughly `2^512` recursive copies before it can return; capture can peg CPU and exhaust memory instead of letting posthog-ios reject the invalid property. Reproduction: reproduced — a focused `iosSimulatorArm64Test` with a list containing itself twice remained at approximately 100% CPU and 1.5 GB RSS for over seven minutes until killed.

File: .changeset/ios-boolean-properties.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"posthog-kmp": patch
+---
+
+Fix boolean event properties being sent as `1`/`0` instead of `true`/`false` on iOS — existing insights and filters written against the numeric values need updating.
```
Comment (marandaneto) [inline .changeset/ios-boolean-properties.md]: suggestion: Document a compatible filter migration — Since historical iOS events remain `1`/`0` while new events use `true`/`false`, please tell users to use an OR condition during the transition (for example, `true OR 1` and `false OR 0`) so events from both older and newer SDK versions continue to match. Reproduction: not reproduced — this is documentation-only compatibility guidance based on the intentional wire-format change described here.

**Review by marandaneto — COMMENTED** [review]
Automated advisory code review.

**Review by marandaneto — APPROVED** [review]
approving to unblock, see comments

File: posthog-kmp/src/iosMain/kotlin/com/posthog/kmp/NativeProperties.ios.kt (hunk 1 above)
Comment (turnipdabeets) [reply posthog-kmp/src/iosMain/kotlin/com/posthog/kmp/NativeProperties.ios.kt]: Addressed in 4ba7206 — replaced the depth-only guard with an ancestor identity stack; a repeat is handed back uncopied like the depth bound already did. Regression test does not finish within 300s against the pre-fix source.

File: .changeset/ios-boolean-properties.md (hunk 2 above)
Comment (turnipdabeets) [reply .changeset/ios-boolean-properties.md]: Addressed in 17e6c98 — the change intent now says filters should match both values until the old data ages out.
